fetchFromGitLab: support for private repositories

panicgh opened the pull request:
###### Description of changes

This is an attempt to add support for private GitLab repositories (via password/token login), similar to `fetchFromGitHub`. I tried to maintain a similar structure as for GitHub, but some details in GitLab API seem to work differently: the `.netrc` file can only be used for the `fetchgit` approach; a custom header to curl needs to be used for `fetchzip` instead ([API doc](https://docs.gitlab.com/ee/api/repositories.html#get-file-archive))

Let me know if it would be better to rebase on `staging` instead of `master`.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

panicgh commented on pkgs/build-support/fetchgitlab/default.nix:
```diff
@@ -1,32 +1,57 @@
+                password ''$${varBase}PASSWORD
+        EOF
+      '' else ''
+        touch netrc
```
This is necessary because `fetchurl` always adds a `--netrc-file` argument if a `netrcPhase` is present, see https://github.com/NixOS/nixpkgs/blob/265ca0dee9c0a200f26ab32bc217055ff8096378/pkgs/build-support/fetchurl/default.nix#L158-L161

panicgh replied:
Added a clarifying comment directly into the source.

panicgh commented on pkgs/build-support/fetchgitlab/default.nix:
```diff
@@ -1,32 +1,57 @@
+        EOF
+      '' else ''
+        touch netrc
+        curlOpts="$curlOpts --header PRIVATE-TOKEN:''$${varBase}PASSWORD"
```
In principle, there should be a space between the header name and argument, but this would require curlOpts to be an array as in #164662.

```suggestion
        curlOptsList+=("--header" "PRIVATE-TOKEN: ''$${varBase}PASSWORD")
```

panicgh replied:
Implemented with a temporary file and `@filename` syntax, which is anyway preferred to avoid leaking the token on the command line.

panicgh commented:
I successfully tested the fetcher with a private GitLab instance and an access token for the "Reporter" role. If `fetchgit` is used (e.g. via `forceFetchGit=true;`), the token needs `read_repository` scope; otherwise (i.e. `fetchzip`), it needs `read_api` scope.

SuperSandro2000 commented on pkgs/build-support/fetchgitlab/default.nix:
```diff
@@ -1,32 +1,62 @@
+  passthruAttrs = removeAttrs args [ "protocol" "domain" "owner" "group" "repo" "rev" "forceFetchGit" "private" "varPrefix" ];
 
-  useFetchGit = deepClone || fetchSubmodules || leaveDotGit;
+  varBase = "NIX${if varPrefix == null then "" else "_${varPrefix}"}_GITLAB_PRIVATE_";
```
```suggestion
  varBase = "NIX${lib.optionalString (varPrefix != null) "_${varPrefix}"}_GITLAB_PRIVATE_";
```

SuperSandro2000 commented on pkgs/build-support/fetchgitlab/default.nix:
```diff
@@ -1,32 +1,62 @@
-, fetchSubmodules ? false, leaveDotGit ? false, deepClone ? false
+, fetchSubmodules ? false, leaveDotGit ? false
+, deepClone ? false, forceFetchGit ? false
+, sparseCheckout ? ""
```
This is a boolean option?

panicgh replied:
I don't think so. It looks rather like a string with directories/files in it:

https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/fetchgit/tests.nix#L15
https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/fetchgit/nix-prefetch-git#L51

SuperSandro2000 replied:
Then it should really be a list of strings rather than a multiline string.

panicgh replied:
In principle I agree, but it would require a (breaking) change in [fetchgit](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/fetchgit/default.nix#L18), [fetchFromGitHub](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/fetchgithub/default.nix#L6), [documentation](https://nixos.org/manual/nixpkgs/stable/#fetchgit), and in some nixpkgs (although not so many and quite simple cases):

```console
$ git grep -l sparseCheckout
doc/builders/fetchers.chapter.md
pkgs/applications/misc/nerd-font-patcher/default.nix
pkgs/build-support/fetchgit/builder.sh
pkgs/build-support/fetchgit/default.nix
pkgs/build-support/fetchgit/nix-prefetch-git
pkgs/build-support/fetchgit/tests.nix
pkgs/build-support/fetchgithub/default.nix
pkgs/build-support/fetchgitlab/default.nix
pkgs/data/fonts/noto-fonts/default.nix
pkgs/data/icons/beauty-line-icon-theme/default.nix
pkgs/data/misc/colemak-dh/default.nix
pkgs/games/itch/default.nix
```

panicgh replied:
As #200082 was finally merged, I changed `sparseCheckout` to a list (of strings).

panicgh commented:
Hmm. I just discovered #43583 and wonder if it still makes sense to continue with this PR.

panicgh committed:
fetchgitlab: pass private tokens/passwords

Allow the fetcher to authenticate with user name and password.
Note: as of now, GitLab ignores the user name (it must be set, but the
value does not matter) but this may change in the future:
https://gitlab.com/gitlab-org/gitlab/-/issues/212953

Credentials can be passed to the nix-daemon, for example, via a
read-protected `EnvironmentFile`:

```console
$ ls -l /to/secrets.txt
-rw------- 1 root root 100 Nov  1 10:42 /to/secrets.txt
```

In /to/secrets.txt:

```
 # for `fetchFromGitLab { private=true; ... }`
NIX_GITLAB_PRIVATE_USERNAME=whatever
NIX_GITLAB_PRIVATE_PASSWORD=glpat-the-access-token
 # for `fetchFromGitLab { private=true; varPrefix="EXAMPLE"; ... }`
NIX_EXAMPLE_GITLAB_PRIVATE_USERNAME=whatever
NIX_EXAMPLE_GITLAB_PRIVATE_PASSWORD=glpat-another-access-token
```

In /etc/nixos/configuration.nix:

```nix
{ config, pkgs, ... }:
{
  systemd.services.nix-daemon.serviceConfig.EnvironmentFile =
    "/to/secrets.txt";
}
```

GitLab supports HTTP Basic Authentication (credentials in `.netrc` file)
only if accessed via Git. Access via the GitLab API requires a custom
header (e.g. `PRIVATE-TOKEN`) instead. See:

* https://docs.gitlab.com/ee/user/project/settings/project_access_tokens.html#project-access-tokens
* https://docs.gitlab.com/ee/api/rest/authentication.html#personalprojectgroup-access-tokens

panicgh commented:
* Rebased to master
* Changed `sparseCheckout` from string to list

Questions:
* Are tests required? How to implement them best? (e.g. private repo with public access token at gitlab.com, ad-hoc test VMs with preconfigured gitlab instance?)
* Is documentation required? (But there is no documentation on the private repo support for fetchFromGitHub either)

panicgh commented on pkgs/build-support/fetchzip/default.nix:
```diff
@@ -22,7 +22,7 @@
 
 
-lib.warnIf (extraPostFetch != "") "use 'postFetch' instead of 'extraPostFetch' with 'fetchzip' and 'fetchFromGitHub'."
+lib.warnIf (extraPostFetch != "") "use 'postFetch' instead of 'extraPostFetch' with 'fetchzip' and 'fetchFromGitHub' or 'fetchFromGitLab'."
```
Squashed into the first commit, together with aligning args to fetchFromGitHub.

SuperSandro2000 commented:
> Are tests required? How to implement them best? (e.g. private repo with public access token at gitlab.com, ad-hoc test VMs with preconfigured gitlab instance?)

We don't have tests for most other more complex fetchers and I think a VM test is overkill.

> Is documentation required? (But there is no documentation on the private repo support for fetchFromGitHub either)

It would be nice to have it otherwise other people don't know how to use it.

SuperSandro2000 commented on pkgs/build-support/fetchgitlab/default.nix:
```diff
@@ -1,32 +1,62 @@
 
-  useFetchGit = deepClone || fetchSubmodules || leaveDotGit;
+  varBase = "NIX${lib.optionalString (varPrefix != null) "_${varPrefix}"}_GITLAB_PRIVATE_";
+  useFetchGit = fetchSubmodules || leaveDotGit || deepClone || forceFetchGit || !(sparseCheckout == [] || sparseCheckout == "");
```
Please don't add support for multiline sparseCheckout which got recently removed.

panicgh replied:
Removed.

SuperSandro2000 commented on pkgs/build-support/fetchgitlab/default.nix:
```diff
@@ -1,32 +1,62 @@
+        # needed because fetchurl always sets --netrc-file if a netrcPhase is present
+        touch netrc
+
+        cat > private-token <<EOF
+        PRIVATE-TOKEN: ''$${varBase}PASSWORD
+        EOF
+        curlOpts="$curlOpts --header @./private-token"
```
Can we not express this in a netrc file?

panicgh replied:
GitLab can use HTTP Basic Authentication and a `.netrc` file *only* when used via Git (e.g. `forceFetchGit = true`).

https://docs.gitlab.com/ee/user/project/settings/project_access_tokens.html#project-access-tokens

> Use a project access token to authenticate:
>
>  *  With the [GitLab API](https://docs.gitlab.com/ee/api/index.html#personalprojectgroup-access-tokens).
>  *  With Git, when using HTTP Basic Authentication, use:
>     *   Any non-blank value as a username.
>     *   The project access token as the password.
>
> Project access tokens are similar to [group access tokens](https://docs.gitlab.com/ee/user/group/settings/group_access_tokens.html) and [personal access tokens](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html).

It needs a custom `PRIVATE-TOKEN` header for access via the GitLab API (`fetchzip`/`fetchurl`):
https://docs.gitlab.com/ee/api/#personalprojectgroup-access-tokens

I added a short clarification to the code and to the commit message.

panicgh commented:
If I manage to do in not too long time, I will also try to submit a test based on the [GitLab NixOS test](https://github.com/NixOS/nixpkgs/blob/e5378ad184e784d1b2348fa777933a1f69725a07/nixos/tests/gitlab.nix).

panicgh commented:
I first rebased and then slightly reworded the documentation and added the NixOS test (based on the gitlab NixOS test). Please review.

roberth commented:
I would generally recommend using the built-in fetcher, because it is designed not to leak the credentials.
On Hercules CI the built-in fetcher even uses a transient token.
That said, this PR looks like a good and well tested addition for those who need the fixed-output derivation approach instead.

roberth commented on nixos/tests/gitlab-fetch-private.nix:
```diff
@@ -0,0 +1,371 @@
+        gitlab.succeed(
+            "${pkgs.sudo}/bin/sudo -u gitlab -H gitlab-rake gitlab:check 1>&2"
+        )
+        gitlab.succeed(
```
I would prefer for the client to be in a separate node. It's just slightly more true to real use.

panicgh commented:
As suggested, I changed the NixOS test to use two VMs. Also I clarified in the documentation that credentials would be visible to all users if used on a multi-user system (`nix-daemon`), which is a clear risk with this approach.

SuperSandro2000 commented on doc/builders/fetchers.chapter.md:
```diff
@@ -137,6 +137,8 @@ A number of fetcher functions wrap part of `fetchurl` and `fetchzip`. They are m
 
 This is used with GitLab repositories. The arguments expected are very similar to `fetchFromGitHub` above.
 
+It is possible to access private repositories by passing the argument `private = true`, and optionally setting a `varPrefix`. This makes `fetchFromGitLab` expect credentials in environment variables in the Nix building process (nix-daemon in multi-user mode). If `varPrefix` is unset, these are `NIX_GITLAB_PRIVATE_USERNAME` and `NIX_GITLAB_PRIVATE_PASSWORD`, and `NIX_MYPREFIX_GITLAB_PRIVATE_USERNA ...
```
```suggestion
It is possible to access private repositories by passing the argument `private = true`, and optionally setting a `varPrefix`. This makes `fetchFromGitLab` expect credentials in environment variables in the Nix building process (nix-daemon in multi-user mode).
If `varPrefix` is unset, these are `NIX_GITLAB_PRIVATE_USERNAME` and `NIX_GITLAB_PRIVATE_PASSWORD`, and `NIX_MYPREFIX_GITLAB_PRIVATE_USERNAME` and `NIX_MYPREFIX_GITLAB_PRIVATE_PASSWORD` if `varPrefix = "MYPREFIX"`.
This allows you to use different credentials for different repositories or groups thereof.
As of now, GitLab [accepts](https://gitlab.com/gitlab-org/gitlab/-/issues/212953) any non-empty user name. The personal/project/group access token must be configured for the *Reporter* role or higher (the *Guest* role works only in [some cases](https://docs.gitlab.com/ee/user/permissions.html#project-members-permissions)) with the `read_api` [scope](https://docs.gitlab.com/ee/user/project/settings/project_access_tokens.html#scopes-for-a-project-access-token) if `fetchzip` is used as the underlying fetcher, or with the `read_repository` scope if the underlying fetcher is `fetchgit`.
Use both scopes simultaneously if any fetcher shall be usable. In multi-user mode, the credentials can be passed to the nix-daemon service via an `EnvironmentFile`. Note however that this makes the credentials available to all users on the system.
```

panicgh replied:
Changed, no problem. But because of Markdown, the rendered output is the same: it is one long paragraph. If you intended to have separate smaller paragraphs, I can change it again. If it is fine as is, please resolve this conversation.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/unable-to-use-fetchgit-on-private-git-repo/24280/5

panicgh commented:
- Rebased to `master`.
- Resolved a warning about deprecated `nodes.gitlab.config.services` -> `nodes.gitlab.services`.
- Backported the fix from 3cd1bd588842b11521ecc7950dad884137b6c97b regarding password complexity.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/using-a-private-gitlab-instance-with-fetchgit/28013/4

roberth commented on nixos/tests/gitlab-fetch-private.nix:
```diff
@@ -0,0 +1,374 @@
+      # Fetch as if the repo was public. This must fail.
+      fetchAsPublic = pkgs.writeText "fetch-as-public.nix" ''
+        let
+          pkgs = import <nixpkgs> {};
```
We could save a lot of rebuilds of this test by pulling the evaluation of this expression out of the VM and into the main evaluation. Instead of using it in the main build though, we can push the `.drv` into the VM. That way we don't need `<nixpkgs>` in the VM, and we won't need to rebuild this test for every little unrelated change in the repo.

So you would write the example as if you're going to use it, but instead of referencing its outputs, you reference its `drvPath` and use `builtins.unsafeDiscardOutputDependency` to remove the output from the string context (it's odd that this step is necessary; wip. It is also not unsafe.)

Instead of `nix-build ${someFile}` you could use the low level `nix-store --realise` command:

    nix-store --realise ${builtins.unsafeDiscardOutputDependency example.drvPath}

panicgh replied:
I tried to implement your suggestion, but got "path ... is not valid" errors. Could you give me more hints what to do? My current attempt is in branch https://github.com/panicgh/nixpkgs/commits/fetchgitlab-netrc-failed_attempt

roberth replied:
Could be either
- nix 2.18.0, a bad release
- bug https://github.com/NixOS/nix/issues/9146

I think it's the latter, because `virtualisation.useNixStoreImage = true;` uses `exportReferencesGraph` for image building.
Could you try with `false`?
Performance difference should not be as dramatic anymore since a performance bug in qemu and/or linux (not sure) was solved. Writable store doesn't need it either.

panicgh replied:
I tried with nix (2.13.5) and nixUnstable (2.15.2) and set `useNixStoreImage=false`. I also tried with and without `invalidateFetcherByDrvHash`, but still get `error: path '/nix/store/s1bf24apdvm6pj79bxibyymvrzhy3bmk-source' is not valid`.

roberth reviewed (commented):
Haven't _validated_ the solution, but I'm inclined to merge this as it is well tested and lgtm.
I'd recommend eval-time fetching, but both solutions have their own merit; this one is faster for package sources, whereas the `builtins.fetch*` is faster for expressions and is a bit easier to set up.
Coming to think of it, it would be great to document the trade-off.

panicgh commented:
Rebased and added an `expires_at` attribute to the API calls creating new tokens (otherwise the call fails). I was unsuccessful using `nix-store --realise` as suggested by @roberth (see discussion above).

If you think it makes sense, I could split out the first 3 commits (to add `forceFetchGit` and `sparseCheckout` to `fetchFromGitLab`) into a separate PR, which is less complex and more generally useful than this one.

panicgh commented:
Rebased. Omitted the commits already upstreamed in #268304.

viraptor commented:
It would be great if you could include an example of usage in practice. I've seen a linked example of nix-daemon environment variables file already, but some sample of non-daemon usage would be useful too. I'm going through a number of ways to try now, so I suspect questions like that will be raised later.

fliiiix commented:
Is there a reason this pr is still in draft?

centromere commented:
@SuperSandro2000 Do you know why this hasn't been merged yet?

panicgh commented:
> Is there a reason this pr is still in draft?

I wanted to rebase it and align with the latest changes from the gitlab NixOS tests, but did not yet have the time to do so. And it seemed I would be the only one interested in this and have found a solution (apply the patch locally) that works for me for the past several years. But if now other people may want to use this, I can do the rebase/potential cleanup etc. next week or so.

Jappie3 commented on pkgs/build-support/fetchgitlab/default.nix:
```diff
@@ -6,18 +6,54 @@ lib.makeOverridable (
+  privateAttrs = lib.optionalAttrs private (
+    lib.throwIfNot (protocol == "https")
+    "private token login is only supported for https" {
+      netrcPhase = ''
```
According to my testing, access to the Gitlab API only requires an access token - not a username and password. I suggest modifying the `if` statement like this:

```nix
netrcImpureEnvVars = ["${varBase}USERNAME" "${varBase}PASSWORD" "${varBase}TOKEN"];
netrcPhase =
  if useFetchGit
  then
    # GitLab supports HTTP Basic Authentication only when Git is used:
    # https://docs.gitlab.com/ee/user/project/settings/project_access_tokens.html#project-access-tokens
    ''
      if [ -z "''$${varBase}USERNAME" -o -z "''$${varBase}PASSWORD" ]; then
        echo "Error: Private fetchFromGitLab using Git requires the nix building process (nix-daemon in multi user mode) to have the ${varBase}USERNAME and ${varBase}PASSWORD env vars set." >&2
        exit 1
      fi

      cat > netrc <<EOF
      machine ${domain}
              login ''$${varBase}USERNAME
              password ''$${varBase}PASSWORD
      EOF
    ''
  else
    # Access via the GitLab API requires a custom header and does not work
    # with HTTP Basic Authentication:
    # https://docs.gitlab.com/ee/api/#personalprojectgroup-access-tokens
    ''
      if [ -z "''$${varBase}TOKEN" ]; then
        echo "Error: Private fetchFromGitLab using the GitLab API requires the nix building process (nix-daemon in multi user mode) to have the ${varBase}TOKEN env var set." >&2
        exit 1
      fi

      # needed because fetchurl always sets --netrc-file if a netrcPhase is present
      touch netrc

      cat > private-token <<EOF
      PRIVATE-TOKEN: ''$${varBase}TOKEN
      EOF
      curlOpts="$curlOpts --header @./private-token"
    '';
```

panicgh replied:
I have no strong argument, but would tend to keep username/password for git and API access. Using the same variables makes it easier to for example change for `forceFetchGit = true` without changing credentials and also makes the fetcher more similar to fetchFromGitHub. Even if the user name is technically unnecessary, maybe it can still be useful for example to print it in logs (I did not verify if this is actually possible).

viraptor commented:
> But if now other people may want to use this

Definitely still interested here.

stuart-warren commented:
There is an upstream issue to support basic auth that could always do with more thumbs up to get it prioritised https://gitlab.com/gitlab-org/gitlab/-/issues/341157

panicgh commented:
Rebased. It seems to work fine when I use this on NixOS, however the VM tests currently do not pass and nix-daemon coredumps in one of the VMs... Maybe someone has an idea?

It seems to fail on the first test "fetch-as-public" (which tries to download a private repo without token, which is expected to fail due to the missing of token). This seems to work in principle (curl errors with 404), but then nix-daemon crashes with "FATAL: exception not rethrown".

```console
$ nix build -f . nixosTests.gitlab-fetch-private -L --no-link  --print-out-paths
...
vm-test-run-gitlab-fetch-private> client # trying https://acme.test/api/v4/projects/alice%2Ftest-alice/repository/archive.tar.gz?sha=refs%2Fheads%2Fmaster
vm-test-run-gitlab-fetch-private> client #   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
vm-test-run-gitlab-fetch-private> client #                                  Dload  Upload   Total   Spent    Left  Speed
vm-test-run-gitlab-fetch-private> gitlab # [  115.759861] gitlab-workhorse[1191]: acme.test 127.0.0.1 - - [2025/01/24:23:15:41 +0000] "GET /api/v4/projects/alice%2Ftest-alice/repository/archive.tar.gz?sha=refs%2Fheads%2Fmaster HTTP/1.1" 404 35 "" "curl/8.11.1 Nixpkgs/25.05" 80
vm-test-run-gitlab-fetch-private> client #   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
vm-test-run-gitlab-fetch-private> client # curl: (22) The requested URL returned error: 404
vm-test-run-gitlab-fetch-private> client # error: cannot download source-salted-7y4klz60xpdh from any mirror
vm-test-run-gitlab-fetch-private> client # error: builder for '/nix/store/c85yf1ab6v6m68wyx6q6nry2xjfkkd30-source-salted-7y4klz60xpdh.drv' failed with exit code 1;
vm-test-run-gitlab-fetch-private> client #        last 9 log lines:
vm-test-run-gitlab-fetch-private> client #        > error checking the existence of https://tarballs.nixos.org/sha256/sha256-HqozwmyEGzXUlSL6zgInlycalxROnD2QOSPRwkrw8sQ=:
vm-test-run-gitlab-fetch-private> client #        > curl: (6) Could not resolve host: tarballs.nixos.org
vm-test-run-gitlab-fetch-private> client #        >
vm-test-run-gitlab-fetch-private> client #        > trying https://acme.test/api/v4/projects/alice%2Ftest-alice/repository/archive.tar.gz?sha=refs%2Fheads%2Fmaster
vm-test-run-gitlab-fetch-private> client #        >   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
vm-test-run-gitlab-fetch-private> client #        >                                  Dload  Upload   Total   Spent    Left  Speed
vm-test-run-gitlab-fetch-private> client #        >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
vm-test-run-gitlab-fetch-private> client #        > curl: (22) The requested URL returned error: 404
vm-test-run-gitlab-fetch-private> client #        > error: cannot download source-salted-7y4klz60xpdh from any mirror
vm-test-run-gitlab-fetch-private> client #        For full logs, run 'nix-store -l /nix/store/c85yf1ab6v6m68wyx6q6nry2xjfkkd30-source-salted-7y4klz60xpdh.drv'.
vm-test-run-gitlab-fetch-private> client # [  127.221911] nix-daemon[946]: FATAL: exception not rethrown
vm-test-run-gitlab-fetch-private> client # [  127.237442] systemd-coredump[1621]: Process 946 (nix-daemon) of user 0 terminated abnormally with signal 6/ABRT, processing...
```

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2600
